File: compositor/src/main/java/org/westford/compositor/core/RoleVisitor.java
package org.westford.compositor.core;


import org.westford.compositor.wlshell.ShellSurface;

public interface RoleVisitor {

    default void visit(final Role role) {defaultAction(role);}

    default void visit(final PointerDevice pointerDevice) {defaultAction(pointerDevice);}

    default void visit(final ShellSurface shellSurface) {defaultAction(shellSurface);}

    default void visit(final Subsurface subsurface) {defaultAction(subsurface);}

    default void defaultAction(final Role role) {}
}

File: compositor/src/main/java/org/westford/compositor/core/RoleVisitor.kt
package org.westford.compositor.core


import org.westford.compositor.wlshell.ShellSurface

interface RoleVisitor {

    fun visit(role: Role) {
        defaultAction(role)
    }

    fun visit(pointerDevice: PointerDevice) {
        defaultAction(pointerDevice)
    }

    fun visit(shellSurface: ShellSurface) {
        defaultAction(shellSurface)
    }

    open fun visit(subsurface: Subsurface) {
        defaultAction(subsurface)
    }

    open fun defaultAction(role: Role) {}
}
